internal/util/flowgraph/node.go

```diff
@@ -53,26 +53,156 @@ type BaseNode struct {
 	maxParallelism int32
 }
 
+// manage nodeCtx
+type nodeCtxManager struct {
+	inputNodeCtx *nodeCtx
+	closeWg      *sync.WaitGroup
+	checker      *timerecord.GroupChecker
+
+	closeOnce sync.Once
+
+	inputNodeCloseCh chan struct{} // notify input node work to exit
+	ddNodeCloseCh    chan struct{} // notify ddnode and downstream node work to exit
+}
+
+// NewNodeCtxManager init with the inputNode and fg.closeWg
+func NewNodeCtxManager(nodeCtx *nodeCtx, closeWg *sync.WaitGroup) *nodeCtxManager {
+	return &nodeCtxManager{
+		inputNodeCtx:     nodeCtx,
+		closeWg:          closeWg,
+		inputNodeCloseCh: make(chan struct{}),
+		ddNodeCloseCh:    make(chan struct{}),
+	}
+}
+
+// Start invoke Node `Start` method and start a worker goroutine
+func (nodeCtxManager *nodeCtxManager) Start(checker *timerecord.GroupChecker) {
+	// in dmInputNode, message from mq to channel, alloc goroutines
+	// limit the goroutines in other node to prevent huge goroutines numbers
+	nodeCtxManager.closeWg.Add(2)
+	go nodeCtxManager.inputNodeWork(checker)
+	go nodeCtxManager.ddNodeWork(checker)
+}
+
+func (nodeCtxManager *nodeCtxManager) inputNodeWork(checker *timerecord.GroupChecker) {
+	defer nodeCtxManager.closeWg.Done()
+	inputNode := nodeCtxManager.inputNodeCtx
+	for {
+		select {
+		case <-nodeCtxManager.inputNodeCloseCh:
+			// remove input node checker
+			checker.Remove(fmt.Sprintf("nodeCtxTtChecker-%s", inputNode.node.Name()))
+			// inputNode will close all node
+			nodeCtxManager.Close()
+			return
+		// handles node work spinning
+		// 1. collectMessage from upstream or just produce Msg from InputNode
+		// 2. invoke node.Operate
+		// 3. deliver the Operate result to downstream nodes
+		default:
+			// inputs from inputsMessages for Operate
+			var input, output []Msg
+			// inputNode.input not from nodeCtx.inputChannel
+			// the input message decides whether the operate method is executed
+			n := inputNode.node
+			inputNode.blockMutex.RLock()
+			if !n.IsValidInMsg(input) {
+				inputNode.blockMutex.RUnlock()
+				return
+			}
+			output = n.Operate(input)
+			inputNode.blockMutex.RUnlock()
+			// the output decide whether the node should be closed.
+			if isCloseMsg(output) {
+				close(nodeCtxManager.inputNodeCloseCh)
+				if inputNode.inputChannel != nil {
+					close(inputNode.inputChannel)
+				}
+				inputNode.Close()
+			}
+			// deliver to all following flow graph node.
+			inputNode.downstream.inputChannel <- output
```

Comment:
you need to change the inputChannel to a buffered channel, with size at least 1

Reply:
this default size is 16